### ISSUE TITLE
Install fail on python 3.6 windows 10

### ISSUE THREAD

**Issue by namjungsoo** [saghul/pyuv#230]
C:\Users\namjungsoo>pip install pyuv
Collecting pyuv
  Using cached pyuv-1.3.0.tar.gz
Installing collected packages: pyuv
  Running setup.py install for pyuv ... error
    Complete output from command c:\users\namjungsoo\appdata\local\programs\python\python36-32\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\NAMJUN~1\\AppData\\Local\\Temp\\pip-build-igqgsf4z\\pyuv\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\NAMJUN~1\AppData\Local\Temp\pip-1jqs9mgd-record\install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build\lib.win32-3.6
    creating build\lib.win32-3.6\pyuv
    copying pyuv\_version.py -> build\lib.win32-3.6\pyuv
    copying pyuv\__init__.py -> build\lib.win32-3.6\pyuv
    running build_ext
    libuv needs to be compiled.
    Building libuv...
    error: No appropriate Python version found. An installation of 2.7 is required to build libuv. You can set the environment variable "PYTHON" to point to a custom installation location.

    ----------------------------------------
Command "c:\users\namjungsoo\appdata\local\programs\python\python36-32\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\NAMJUN~1\\AppData\\Local\\Temp\\pip-build-igqgsf4z\\pyuv\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\NAMJUN~1\AppData\Local\Temp\pip-1jqs9mgd-record\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\NAMJUN~1\AppData\Local\Temp\pip-build-igqgsf4z\pyuv\

**Comment by saghul** [comment]
> error: No appropriate Python version found. An installation of 2.7 is required to build libuv. You can set the environment variable "PYTHON" to point to a custom installation location.

Do you have Python 2.7 installed as well? At the moment it's needed even if you are installing libuv for Python 3.  It sucks, but this is because gyp only supports Python 2.

**Comment by fyears** [comment]
@namjungsoo 

1. download the latest repo
2. install a separated python 2.7.
3. open https://github.com/saghul/pyuv/blob/v1.x/setup_libuv.py#L110, go to line 110. It's the line determining the path to `python.exe` (the specially installed python 2.7). Change it according to you configuration.
4. Run `python setup.py install`.

It should be fine to compile and install.

@saghul 

However, I find some errors running the tests on Python 3.6 on Windows 10:

```bash
$ python -m unittest discover pyuv/tests/

======================================================================
ERROR: test_process_args (test_process.ProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_process.py", line 154, in test_process_args
    self.loop.run()
  File "pyuv\tests\common.py", line 51, in run
    self._reraise()
  File "pyuv\tests\common.py", line 62, in _reraise
    reraise(typ, value, tb)
  File "pyuv\tests\common.py", line 19, in reraise
    raise value.with_traceback(tb)
  File "pyuv\tests\test_process.py", line 143, in stdout_read_cb
    self.received_output = data.strip()
AttributeError: 'NoneType' object has no attribute 'strip'

======================================================================
ERROR: test_process_env (test_process.ProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_process.py", line 182, in test_process_env
    self.loop.run()
  File "pyuv\tests\common.py", line 51, in run
    self._reraise()
  File "pyuv\tests\common.py", line 62, in _reraise
    reraise(typ, value, tb)
  File "pyuv\tests\common.py", line 19, in reraise
    raise value.with_traceback(tb)
  File "pyuv\tests\test_process.py", line 170, in stdout_read_cb
    self.received_output = data.strip()
AttributeError: 'NoneType' object has no attribute 'strip'

======================================================================
ERROR: test_process_kill (test_process.ProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_process.py", line 242, in test_process_kill
    self.loop.run()
  File "pyuv\tests\common.py", line 51, in run
    self._reraise()
  File "pyuv\tests\common.py", line 62, in _reraise
    reraise(typ, value, tb)
  File "pyuv\tests\common.py", line 19, in reraise
    raise value.with_traceback(tb)
  File "pyuv\tests\test_process.py", line 236, in timer_cb
    proc.kill(15)
pyuv._cpyuv.error.HandleClosedError: Handle is closing/closed

======================================================================
ERROR: test_open (test_tcp.TCPErrorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_tcp.py", line 38, in test_open
    sock.close()
  File "d:\prog\anaconda\Lib\socket.py", line 417, in close
    self._real_close()
  File "d:\prog\anaconda\Lib\socket.py", line 411, in _real_close
    _ss.close(self)
OSError: [WinError 10038] An operation was attempted on something that is not a socket

======================================================================
FAIL: test_process_basic (test_process.ProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_process.py", line 29, in test_process_basic
    self.loop.run()
  File "pyuv\tests\common.py", line 51, in run
    self._reraise()
  File "pyuv\tests\common.py", line 62, in _reraise
    reraise(typ, value, tb)
  File "pyuv\tests\common.py", line 19, in reraise
    raise value.with_traceback(tb)
  File "pyuv\tests\test_process.py", line 22, in proc_exit_cb
    self.assertEqual(exit_status, 0)
AssertionError: 2 != 0

======================================================================
FAIL: test_process_cwd (test_process.ProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_process.py", line 100, in test_process_cwd
    self.loop.run()
  File "pyuv\tests\common.py", line 51, in run
    self._reraise()
  File "pyuv\tests\common.py", line 62, in _reraise
    reraise(typ, value, tb)
  File "pyuv\tests\common.py", line 19, in reraise
    raise value.with_traceback(tb)
  File "pyuv\tests\test_process.py", line 93, in proc_exit_cb
    self.assertEqual(exit_status, 0)
AssertionError: 2 != 0

======================================================================
FAIL: test_process_stdin (test_process.ProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_process.py", line 220, in test_process_stdin
    self.assertEqual(self.received_output, b"TEST")
AssertionError: None != b'TEST'

======================================================================
FAIL: test_process_stdout (test_process.ProcessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pyuv\tests\test_process.py", line 127, in test_process_stdout
    self.loop.run()
  File "pyuv\tests\common.py", line 51, in run
    self._reraise()
  File "pyuv\tests\common.py", line 62, in _reraise
    reraise(typ, value, tb)
  File "pyuv\tests\common.py", line 19, in reraise
    raise value.with_traceback(tb)
  File "pyuv\tests\test_process.py", line 111, in proc_exit_cb
    self.assertEqual(exit_status, 0)
AssertionError: 2 != 0

----------------------------------------------------------------------
Ran 232 tests in 9.657s

FAILED (failures=4, errors=4, skipped=26)
```

**Comment by fischmax** [comment]
I installed python 2.7, made sure it was in my PATH and tried again

    pip3 install --user pyuv

What I got was

```
Installing collected packages: pyuv
  Running setup.py install for pyuv: started
    Running setup.py install for pyuv: finished with status 'error'
    Complete output from command c:\users\maxi\appdata\local\programs\python\python36-32\python.exe -u -c "import setuptools, tokenize;__file__='C:\\U                                                                              sers\\Maxi\\AppData\\Local\\Temp\\pip-build-ebzndrim\\pyuv\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');                                                                              f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\Maxi\AppData\Local\Temp\pip-ri_3s1mv-record\install-record.txt --single-version-externally-managed --compile --user --prefix=:
    running install
    running build
    running build_py
    creating build
    creating build\lib.win32-3.6
    creating build\lib.win32-3.6\pyuv
    copying pyuv\_version.py -> build\lib.win32-3.6\pyuv
    copying pyuv\__init__.py -> build\lib.win32-3.6\pyuv
    running build_ext
    libuv needs to be compiled.
    Building libuv...
    Using "C:\Python27\python.exe" to build libuv...
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\Maxi\AppData\Local\Temp\pip-build-ebzndrim\pyuv\setup.py", line 43, in <module>
        sources = ['src/pyuv.c'],
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\site-packages\setuptools\command\install.py", line 61, in run
        return orig.install.run(self)
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\command\install.py", line 545, in run
        self.run_command('build')
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\command\build.py", line 135, in run
        self.run_command(cmd_name)
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\distutils\command\build_ext.py", line 339, in run
        self.build_extensions()
      File "C:\Users\Maxi\AppData\Local\Temp\pip-build-ebzndrim\pyuv\setup_libuv.py", line 158, in build_extensions
        self.get_libuv()
      File "C:\Users\Maxi\AppData\Local\Temp\pip-build-ebzndrim\pyuv\setup_libuv.py", line 225, in get_libuv
        build_libuv()
      File "C:\Users\Maxi\AppData\Local\Temp\pip-build-ebzndrim\pyuv\setup_libuv.py", line 198, in build_libuv
        exec_process(['cmd.exe', '/C', 'vcbuild.bat', libuv_arch, 'release'], cwd=self.libuv_dir, env=env, shell=True, silent=not self.libuv_verbose_build)
      File "C:\Users\Maxi\AppData\Local\Temp\pip-build-ebzndrim\pyuv\setup_libuv.py", line 50, in exec_process
        stdout = stdout.decode(sys.stdout.encoding)
      File "c:\users\maxi\appdata\local\programs\python\python36-32\lib\encodings\cp1252.py", line 15, in decode
        return codecs.charmap_decode(input,errors,decoding_table)
    UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 288: character maps to <undefined>

    ----------------------------------------
Command "c:\users\maxi\appdata\local\programs\python\python36-32\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\Maxi\\AppData\\Local\\Temp\\pip-build-ebzndrim\\pyuv\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\Maxi\AppData\Local\Temp\pip-ri_3s1mv-record\install-record.txt --single-version-externally-managed --compile --user --prefix=" failed with error code 1 in C:\Users\Maxi\AppData\Local\Temp\pip-build-ebzndrim\pyuv\
```

which seems to be an older issue. However, I do not have non-ascii characters in my PATH, working directory or username. Heck, I even avoid Spaces in these places. Right now I am already way too deep in the rabbit hole of arcane python configurations. Send help!

**Comment by schlamar** [comment]
@saghul Uploading wheels for 3.6 to PyPI should resolve most of these issues :)

**Comment by saghul** [comment]
@schlamar Definitely! I see you opened a PR for 3.6 on AppVeyor, cheers for that!

**Comment by schlamar** [comment]
@fischmax UnicodeError was fixed #239 

@fyears the error in test_tcp will be fixed in #235. I cannot reproduce the other issues, probably they are already fixed in the v1.x branch.

**Comment by schlamar** [comment]
@fyears Please create a new issue if you see other test errors than the one in test_tcp on the v1.x branch. BTW, instead of manipulating setup_libuv.py, you can just point set the `PYTHON` environment variable to your Python 2.7 executable. However, this should not be necessary if Python is installed in the default location (C:\Python27).
